active/0021-transfer-files-over-mqtt.md

```diff
@@ -0,0 +1,206 @@
+* Client uses $file Topic Filter to transfer files
+* Clients cannot subscribe to $file topics
+* Segment length can be calculated on the server side by subtracting the length of the [Variable Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901025) from the [Remaining Length](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901105) field that is in the [Fixed Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901021)
+* Data is transferred in PUBLISH packets in the following order:
+  1. $file/{fileId}/init
+  2. $file/{fileId}/{offset}/{sha256sum}
+  3. $file/{fileId}/{offset}/{sha256sum}
+  4. ...
+  3. $file/{fileId}/[fin/{sha256sum} | abort]
+* Client can send up to N (configured) PUBLISH packets before blocking for PUBACK
+
+#### `$file/{fileId}/init` message
+
+Initialize the file transfer. Server is expected to store metadata from the payload in the session along with `{fileId}` as a reference for the rest of file metadata.
+
+  * Qos=1
+  * DUP=0 for the initial segment transfer, 1 when retransmitting
+  * Payload Format Indicator=0x01
+  * `{fileId}` is corresponding file UUID
+  * Payload is a JSON document with the following properties:
+    * filepath
+      * a character string composed with characters from this set [a-zA-Z0-9_-./]
+      * length in bytes MUST be <= 4096 (PATH_MAX=4096 on most common Linux filesystems)
+    * filename
+      * a character string composed with characters from this set [a-zA-Z0-9_-./]
+      * length in bytes MUST be < 256 (NAME_MAX=255 in Linux)
+    * size [optional]
+      * total file size in bytes
+
+#### `$file/{fileId}/{offset}/{sha256sum}` message
+
+One such message for each file segment.
+
+  * Qos=1
+  * DUP=0 for the initial segment transfer, 1 when retransmitting
+  * Payload Format Indicator=0x00
+  * Packet Identifier=sequential file segment number
+  * Payload is file segment bytes
+  * `{offset}` is byte offset of the given segment
+  * `{sha256sum}` is sha256 checksum of file segment bytes
+
+#### `$file/{fileId}/fin/{sha256sum}` message
+
+All file segments have been successfully transferred.
+
+  * Qos=1
+  * no payload
+  * `{sha256sum}` is sha256 checksum of the file
+
+#### `$file/{fileId}/abort` message
+
+Client wants to abort the transfer.
+
+  * Qos=1
+  * no payload
+
+### PUBACK Reason codes
+
+| Reason code | MQTT Name                     | Meaning in file transfer context                    |
+|-------------|-------------------------------|-----------------------------------------------------|
+| OMIT        |                               | Same as 0x00                                        |
+| 0x00        | Success                       | File segment has been successfully persisted        |
+| 0x10        | No matching subscribers       | Server asks Client to retransmit all segments       |
```

Comment:
* could be used when e.g. file retention kicks in while client is still trying to send some later segments